File: apps/dotnet/src/MachineLearning/Chapter08/Main.fs
module MachineLearning.Chapter08.Main

open System.IO
open MachineLearning.Dataset
open MachineLearning.Chapter02.IrisPreprocessing
open MachineLearning.Chapter08.SurvivedData
open MachineLearning.Chapter08.WeightedTree
open MachineLearning.Chapter08.Pipeline
open MachineLearning.Chapter08.ModelFile
open MachineLearning.Chapter08.MlNetAdapter

[<Literal>]
let TestSize = 0.2

[<Literal>]
let Seed = 0

[<Literal>]
let MaxDepth = 5

/// 学習済みのパイプラインとモデルの保存先（apps/dotnet/model/ は .gitignore の対象）
[<Literal>]
let ModelDirectory = "model"

[<Literal>]
let PipelineFileName = "survived.json"

[<Literal>]
let MlNetFileName = "survived-mlnet.zip"

let ClassWeights = [ Unweighted; Balanced ]

/// 年齢が欠けた架空の乗客（1 等客室の女性、3 等客室の男性）
let NewPassengers: Passenger list =
    [
        {
            Pclass = 1
            Sex = "female"
            Age = None
            SibSp = 0
            Parch = 0
            Fare = 50.0
            Embarked = Some "C"
        }
        {
            Pclass = 3
            Sex = "male"
            Age = None
            SibSp = 0
            Parch = 0
            Fare = 8.0
            Embarked = Some "S"
        }
    ]

let private formatList (values: int list) : string =
    values |> List.map string |> String.concat ", " |> sprintf "[%s]"

/// Survived.csv でクラスの重みの有無を比べ、ML.NET と突き合わせ、パイプラインを保存して読み込む
let runWith (modelDirectory: string) (print: string -> unit) : unit =
    let rows = loadSurvived (Path.Combine(dataDir (), "Survived.csv"))
    let x, t = splitFeaturesAndTarget rows
    let split = splitTrainTest TestSize Seed x t
    let survivors = t |> List.filter ((=) 1) |> List.length
    print $"データ件数: {rows.Length}（生存 {survivors}, 死亡 {rows.Length - survivors}）"
    print $"訓練データ: {split.XTrain.Length} 件, テストデータ: {split.XTest.Length} 件"

    let pipelines =
        ClassWeights
        |> List.map (fun classWeight ->
            classWeight,
            fitPipeline
                {
                    MaxDepth = Some MaxDepth
                    ClassWeight = classWeight
                }
                split.XTrain
                split.TTrain)
        |> Map.ofList

    for classWeight in ClassWeights do
        let pipeline = pipelines[classWeight]
        let result = evaluate pipeline split

        let fastTree =
            trainFastTree
                (pown 2 MaxDepth)
                (split.XTrain |> List.map (transform pipeline))
                split.TTrain
                (classWeights classWeight split.TTrain)

        let mlNet = predictFastTree fastTree (split.XTest |> List.map (transform pipeline))

        let agreed =
            List.zip (predict pipeline split.XTest) mlNet
            |> List.filter (fun (a, b) -> a = b)
            |> List.length

        print (
            $"classWeight={classWeight}: 訓練 {result.TrainAccuracy:F3}, テスト {result.TestAccuracy:F3}, "
            + $"生存者 {result.Survivors} 人中 {result.FoundSurvivors} 人を発見, ML.NET と一致 {agreed}/{split.XTest.Length}"
        )

    let balanced = pipelines[Balanced]
    let pipelineFile = Path.Combine(modelDirectory, PipelineFileName)
    saveModel pipelineFile balanced

    match loadModel pipelineFile with
    | Ok loaded -> print $"架空の乗客の予測（{PipelineFileName}）: {formatList (predict loaded NewPassengers)}"
    | Error message -> print $"{PipelineFileName} を読み込めません: {message}"

    let mlNetFile = Path.Combine(modelDirectory, MlNetFileName)

    trainFastTree
        (pown 2 MaxDepth)
        (split.XTrain |> List.map (transform balanced))
        split.TTrain
        (classWeights Balanced split.TTrain)
    |> saveFastTree mlNetFile

    let mlNetPredictions =
        predictFastTree (loadFastTree mlNetFile) (NewPassengers |> List.map (transform balanced))

    print $"架空の乗客の予測（{MlNetFileName}）: {formatList mlNetPredictions}"

let run (print: string -> unit) : unit = runWith ModelDirectory print

File: apps/scala/src/main/scala/machinelearning/chapter08/Main.scala
package machinelearning.chapter08

import java.nio.file.{Path, Paths}
import java.util.Locale
import machinelearning.chapter02.{Preprocessing, Row, Table}
import machinelearning.dataset.DataDir

/** クラスの重みごとの評価結果を表示し、学習済みのパイプラインを保存して読み込む。 */
object Main:
  /** 学習済みのパイプラインの保存先（apps/scala/model/ は .gitignore の対象） */
  val ModelFile: Path = Paths.get("model", "survived.model")

  private val TestSize = 0.2
  private val Seed = 0L
  private val MaxDepth = Some(5)

  /** 年齢が分からない架空の乗客 2 人（1 等客室の女性と、3 等客室の男性） */
  private val NewPassengers: Table =
    SurvivedData.features(
      Vector(
        passenger("1", "female", "", "0", "0", "50", "C"),
        passenger("3", "male", "", "0", "0", "8", "S")
      )
    )

  /** 特徴量の列の順に並べた値から、乗客 1 人分の行を作る。空文字列は欠損値。 */
  private def passenger(values: String*): Row =
    Row(SurvivedData.FeatureColumns.zip(values.toVector).toMap)

  def run(print: String => Unit): Unit = run(print, ModelFile)

  /** 保存先を指定して実行する。テストから一時ディレクトリを渡すために使う。 */
  def run(print: String => Unit, modelFile: Path): Unit =
    val rows = Table.load(Paths.get(DataDir.current(), "Survived.csv")).rows
    val t = SurvivedData.target(rows)
    val split = Preprocessing.splitTrainTest(rows, t, TestSize, Seed)
    val survived = t.count(_ == 1)
    print(s"データ件数: ${rows.size}（生存 $survived, 死亡 ${t.size - survived}）")
    print(s"訓練データ: ${split.xTrain.size} 件, テストデータ: ${split.xTest.size} 件")

    val pipelines = ClassWeight.values.toVector.map { classWeight =>
      val pipeline =
        Pipeline.build(MaxDepth, classWeight).fit(SurvivedData.features(split.xTrain), split.tTrain)
      val result = Evaluation.evaluate(pipeline, split)
      print(
        s"classWeight=${classWeight.label}: 訓練 ${format(result.trainAccuracy)}, " +
          s"テスト ${format(result.testAccuracy)}, " +
          s"生存者 ${result.survivors} 人中 ${result.foundSurvivors} 人を発見"
      )
      classWeight -> pipeline
    }.toMap

    ModelFiles.save(pipelines(ClassWeight.Balanced), modelFile)
    val predictions = ModelFiles.load(modelFile).predict(NewPassengers)
    print(s"保存したモデル: ${modelFile.getFileName}")
    print(s"架空の乗客の予測: $predictions")

  private def format(value: Double): String = String.format(Locale.ROOT, "%.3f", value)
